api/catalog/api/controllers/elasticsearch/utils.py

```diff
@@ -0,0 +1,160 @@
+    """
+    Hide data sources from the catalog dynamically. This excludes providers with
+    ``filter_content`` enabled from the search results.
+
+    :param s: the search query to issue to Elasticsearch
+    :return: the modified search query
+    """
+
+    filter_cache_key = "filtered_providers"
+    filtered_providers = cache.get(key=filter_cache_key)
+    if not filtered_providers:
+        filtered_providers = ContentProvider.objects.filter(filter_content=True).values(
+            "provider_identifier"
+        )
+        cache.set(
+            key=filter_cache_key,
+            timeout=FILTER_CACHE_TIMEOUT,
+            value=filtered_providers,
+        )
+    if len(filtered_providers) != 0:
+        to_exclude = [f["provider_identifier"] for f in filtered_providers]
+        s = s.exclude("terms", provider=to_exclude)
+    return s
+
+
+def _paginate_with_dead_link_mask(
+    s: Search, page_size: int, page: int
+) -> Tuple[int, int]:
+    """
+    Given a query, a page and page_size, return the start and end
+    of the slice of results.
+
+    :param s: The elasticsearch Search object
+    :param page_size: How big the page should be.
+    :param page: The page number.
+    :return: Tuple of start and end.
+    """
+    query_hash = get_query_hash(s)
+    query_mask = get_query_mask(query_hash)
+    if not query_mask:
+        start = 0
+        end = ceil(page_size * page / (1 - DEAD_LINK_RATIO))
+    elif page_size * (page - 1) > sum(query_mask):
+        start = len(query_mask)
+        end = ceil(page_size * page / (1 - DEAD_LINK_RATIO))
+    else:
+        accu_query_mask = list(accumulate(query_mask))
+        start = 0
+        if page > 1:
+            try:
+                start = accu_query_mask.index(page_size * (page - 1) + 1)
+            except ValueError:
+                start = accu_query_mask.index(page_size * (page - 1)) + 1
+        if page_size * page > sum(query_mask):
+            end = ceil(page_size * page / (1 - DEAD_LINK_RATIO))
+        else:
+            end = accu_query_mask.index(page_size * page) + 1
+    return start, end
+
+
+def _get_query_slice(
+    s: Search, page_size: int, page: int, filter_dead: Optional[bool] = False
+) -> Tuple[int, int]:
+    """
+    Select the start and end of the search results for this query.
+    """
+    if filter_dead:
+        start_slice, end_slice = _paginate_with_dead_link_mask(s, page_size, page)
+    else:
+        # Paginate search query.
+        start_slice = page_size * (page - 1)
+        end_slice = page_size * page
+    if start_slice + end_slice > ELASTICSEARCH_MAX_RESULT_WINDOW:
+        raise ValueError("Deep pagination is not allowed.")
+    return start_slice, end_slice
+
+
+def _post_process_results(
+    s, start, end, page_size, search_results, filter_dead
+) -> List[Hit]:
+    """
+    After fetching the search results from the back end, iterate through the
+    results, perform image validation, and route certain thumbnails through our
```

Comment:
This is not really in the scope of this PR, but where exactly is this done: `route certain thumbnails thumbnails through our proxy` in `_post_process_results` function?

Reply:
This code is in `api/utils/validate_images.py`. It's not exactly what the comment describes though, it checks for broken images but it does not route any images through our proxy. I think we used to do that before for `http://` images but I think we stopped doing sometime and didn't update the comment.